### PR TITLE
Added last.fm plugin

### PR DESCRIPTION
**Pull request by k-e-l-p**

### CODE REVIEW

File: plugins/last-fm.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+api_url = 'http://ws.audioscrobbler.com/2.0/?format=json'
+
+
+def _get_lastfm_nick(conn, irc_nick):
+    nicks = db.get_row(conn, 'lastfm', 'irc_nick', irc_nick)
+    try:
+        return nicks[0][1]
+    except IndexError:
+        pass
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: Could just return None here, that then should remove the need for `nick_check = None` on line 48. Also line 51 you could probably do `if not nick_check:`.

File: plugins/last-fm.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+def _pull_latest_track(api_key, lastfm_nick):
+    url = api_url+f'&method=user.getrecenttracks'
+    url += f'&user={lastfm_nick}&api_key={api_key}'
+    response = requests.get(url)
+    if response.status_code == 200:
+        try:
+            track = response.json()['recenttracks']['track'][0]
+            return track
+        except IndexError:
+            pass
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: Again you could just `return None` and remove the `track = None` on line 214.

File: plugins/last-fm.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    if response.status_code == 200:
+        try:
+            tags_json = response.json()['toptags']['tag'][0:5]
+        except IndexError:
+            tags_json = response.json()['toptags']['tag']
+
+        if len(tags_json) < 1:
+            return 'none'
+
+        tags = []
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: `tags = ' , '.join(tag['name'] for tag in tags_json)` should work, or `tags = ' , '.join([tag['name'] for tag in tags_json])` if join doesn't like the generator expression, I forget if it cares.

File: plugins/last-fm.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+            tags.append(tag['name'])
+        tags = ', '.join(tags)
+        return tags
+
+
+def _get_when_played(track):
+    """returns the time difference between right now and the time the
+    track was played in text format"""
+
+    """i decided to not include this in the output since the API
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: That sorta stuff happens with normal taiga too I think, it depends on your scrobbler and the api. Also I'd look into copying the time conversion code stuff from taiga and putting it into a util, it should be in `plugins/util/timeformat.py` or something, or I might get on that soon. Also you can still use fstrings here, `listened = f'{listened[3:-3]} minutes'` for example.

File: plugins/last-fm.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            return str(playcount_json)
+        except:
+            return 'unknown'
+
+
+def _output_nowplaying(client, data, lastfm_nick,
+                       title, artist,album, plays, tags):
+    out = f'{lastfm_nick} is listening to "{title}" by {artist} from '
+    if album == 'an unknown album':
+        out += 'an unknown album'
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: Could use `out = f'{out} an unknown album'`

File: plugins/last-fm.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+    asyncio.create_task(
+        client.message(
+            data.target, out))
+
+    return
+
+
+@hook.hook('command', ['lastfminit'], admin=True)
+async def lfminit(client, data):
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: You don't need a whole table just for the lastfm, you can just add a column to the users table with the lastfm name. `        db.add_column(conn, 'users', 'lastfm')`, you could add this to [admin.py](https://github.com/FrozenPigs/Taigabot3/blob/master/plugins/admin.py#L50) where I got some db stuff for the channels, or in [db.py,](https://github.com/FrozenPigs/Taigabot3/blob/master/core/db.py#L12) or you could just add it at the top of the command or something like that.

File: plugins/last-fm.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    db.ccache()
+    print('Initialization complete.')
+
+
+@hook.hook('command', ['np'])
+async def nowplaying(client, data):
+    """gets the current song that's playing."""
+
+    conn = client.bot.dbs[data.server]
+    split = data.message.split()
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: `data.split_message`, and preffer to call the variable message.

File: plugins/last-fm.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    conn = client.bot.dbs[data.server]
+    print(f'Initializing lastfm table in /persist/db/{data.server}.db...')
+    db.init_table(conn, 'lastfm', quote_columns)
+    db.ccache()
+    print('Initialization complete.')
+
+
+@hook.hook('command', ['np'])
+async def nowplaying(client, data):
+    """gets the current song that's playing."""
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: Docstrings are used for the commands help, see what I've done with the doc strings in admin.py and gadmin.py.

File: plugins/last-fm.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    except KeyError:
+        asyncio.create_task(
+            client.message(
+                data.target,
+                'No last.fm API key found. Please create an API account, add it\'s API key to \'api_keys\' in config.json with the tag \'last_fm\''))
+        return
+
+    # deal with getting the last.fm nickname
+
+    lastfm_nick = None
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: With the change on line 44 I don't think this should be needed.

File: plugins/last-fm.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        lastfm_nick = _get_lastfm_nick(conn, data.nickname)
+
+        if lastfm_nick is None:
+            asyncio.create_task(
+                client.message(
+                    data.target,
+                    'No associated last.fm nickname found.'))
+            return
+
+    track = None
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: Not needed with line 71 change.

File: plugins/last-fm.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+@hook.hook('command', ['np'])
+async def nowplaying(client, data):
+    """gets the current song that's playing."""
+
+    conn = client.bot.dbs[data.server]
+    split = data.message.split()
+
+    try:
+        api_keys = client.bot.config['api_keys']
```
Comment (FrozenPigs) [inline plugins/last-fm.py]: `lastfm_key = client.bot.config['api_keys']['last_fm']`

File: plugins/last-fm.py (hunk 5 above)
Comment (k-e-l-p) [reply plugins/last-fm.py]: fstrings are actually a *tiny* bit slower than just doing out += var, so i see no point in using them unless i need to insert a variable. I know that the difference is minute, but here is some proof:

appending via +=:
```
#test.py
out = ''

for i in range(0, 100000):
    out += 'a'
```
```
$time python test.py
real	0m0.041s
user	0m0.037s
sys	0m0.004s
```
fstrings:
```
#test.py
out = ''

for i in range(0, 100000):
    out = f'{out}a'
```
```
$time python test.py
real	0m0.247s
user	0m0.213s
sys	0m0.034s
```

**Pull request merged**